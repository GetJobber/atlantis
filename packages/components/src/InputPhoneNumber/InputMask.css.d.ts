declare const styles: {
  readonly "mask": string;
};
export = styles;

