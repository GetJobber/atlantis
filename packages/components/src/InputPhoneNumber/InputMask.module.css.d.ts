declare const styles: {
  readonly "mask": string;
  readonly "hiddenValue": string;
  readonly "cursorPosition": string;
};
export = styles;

