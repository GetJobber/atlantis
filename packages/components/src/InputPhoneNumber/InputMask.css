.container {
  position: relative;
}

.mask {
  position: absolute;
  top: 20px;
  left: 17px;
  z-index: var(--elevation-base);
  color: var(--color-grey);
  font-family: var(--typography--fontFamily-normal);
  font-size: var(--typography--fontSize-base);
  pointer-events: none;
}
