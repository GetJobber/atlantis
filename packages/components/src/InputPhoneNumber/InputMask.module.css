.mask {
  display: flex;
  align-items: center;
  position: absolute;
  top: 0;
  right: 0;
  left: 0;
  z-index: var(--elevation-base);
  height: var(--field--height);
  padding-top: var(--field--padding-top);
  padding-bottom: var(--field--padding-bottom);
  padding-left: var(--field--padding-left);
  padding-right: var(--field--padding-right);
  color: var(--color-grey);
  font-family: inherit;
  font-size: inherit;
  line-height: var(--field--value-lineHeight);
  pointer-events: none;
}

.hiddenValue {
  visibility: hidden;
  opacity: 0;
}

input.emptyPhoneNumber {
  padding-left: calc(var(--field--padding-left) + var(--space-smaller));
}
