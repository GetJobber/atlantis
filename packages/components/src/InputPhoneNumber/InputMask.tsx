import React, { ReactElement, cloneElement } from "react";
import styles from "./InputMask.css";
import { FormFieldProps } from "../FormField";

interface InputMaskProps {
  readonly pattern: string;
  readonly delimiter?: string;
  readonly children: ReactElement<FormFieldProps>;
}

export function InputMask({
  pattern,
  delimiter = "*",
  children,
}: InputMaskProps) {
  const { value: inputValue, onChange } = children.props;
  const placeholderValue = pattern
    .replace(new RegExp(`\\${delimiter}`, "g"), "_")
    .slice(String(inputValue).length);

  return (
    <>
      {cloneElement(children, {
        onChange: handleChange,
        children: (
          <div className={styles.mask} aria-hidden="true">
            <span style={{ opacity: 0 }}>{String(inputValue)}</span>
            {placeholderValue}
          </div>
        ),
      })}
    </>
  );

  function handleChange(value: string) {
    const patternChars = pattern.split("");
    const specialChars = patternChars.filter(char => char !== delimiter);
    const cleanVal = value
      .split("")
      .filter(char => !specialChars.includes(char));
    const newMaskedValue = patternChars.reduce((result, item) => {
      if (!cleanVal.length) return result;
      if (specialChars.includes(item)) return result + item;
      if (item === delimiter) return result + cleanVal.shift();
      return result;
    }, "");

    onChange?.(newMaskedValue);
  }
}
