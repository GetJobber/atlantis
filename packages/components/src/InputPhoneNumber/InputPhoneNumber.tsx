import React from "react";
import { FormField, FormFieldProps } from "../FormField";

export interface RegionSettings {
  countryCode?: number;
  format?: string;
}

export interface AllowedRegions {
  NorthAmerica: RegionSettings;
  GreatBritain: RegionSettings;
  Unknown: RegionSettings;
}

const REGION_SETTINGS: AllowedRegions = {
  NorthAmerica: { countryCode: 1, format: "(###) ###-####" },
  GreatBritain: { countryCode: 44 },
  Unknown: {},
};

// should it be form field or common form field?
export interface InputPhoneNumberProps extends FormFieldProps {
  alwaysShowMask?: boolean;
  region?: keyof AllowedRegions;
  placeholder?: string;
  showCountryCode?: boolean;
}

export function InputPhoneNumber({
  region = "NorthAmerica",
  placeholder,
  alwaysShowMask = true,
  showCountryCode = false,
  ...rest
}: InputPhoneNumberProps) {
  const regionSettings = REGION_SETTINGS[region];

  return (
    <FormField
      value={rest.value}
      onChange={rest.onChange}
      type="maskedNumber"
      maskingProperties={getMaskingProperties(regionSettings)}
      placeholder={placeholder}
      {...rest}
    />
  );

  function getDisplayedCountryCode(settings: RegionSettings) {
    if (showCountryCode && settings.countryCode) {
      return `+${settings.countryCode} `;
    }

    return "";
  }

  function getMaskingProperties(
    settings: RegionSettings,
  ): FormFieldProps["maskingProperties"] {
    const hasFormat = !!settings.format;
    const countryCodeString = getDisplayedCountryCode(regionSettings);

    if (!hasFormat) {
      return {
        prefix: countryCodeString,
      };
    }

    return {
      format: `${countryCodeString}${regionSettings.format}`,
      allowEmptyFormatting: alwaysShowMask,
      mask: "_",
    };
  }
}
