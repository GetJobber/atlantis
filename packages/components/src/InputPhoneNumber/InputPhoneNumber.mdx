---
name: InputPhoneNumber
menu: Components
route: /components/input-phone-number
showDirectoryLink: true
---

import { Playground, Props } from "docz";
import { InputPhoneNumber } from ".";
import AllowedCountries from "./InputPhoneNumber";
import { Heading } from "../Heading";

# Input Phone Number

Input phone number should be used anywhere you need to accept phone numbers as
input.

```ts
import { InputPhoneNumber } from "@jobber/components/InputPhoneNumber";
import { Heading } from "@jobber/components/Heading";
```

<Playground>
  <InputPhoneNumber showCountryCode={true} />
</Playground>

## Design Usage & Guidelines

This component is best suited for when you want to ensure the user adhears to a
specific phone number format and present that information in a friendly-easily
consumable way.

## Props

<Props of={InputPhoneNumber} />

## Region Selection

You can select between the countries that are supported.

By default, we will assume `NorthAmerica`

<ul>
  <li> North America </li>
  <li> Great Britain </li>
  <li> Unknown </li>
</ul>

<Playground>
  <InputPhoneNumber region={"GreatBritain"} showCountryCode={true} />
</Playground>

## Always Show Mask

You may not want the mask to appear until a user begins to interact with the
input, in which case you would set this value to false.

By default, we will assume `true`.

<Playground>
  <InputPhoneNumber region={"NorthAmerica"} alwaysShowMask={false} />
</Playground>

## Show Country Code

This switch just hides or shows the country code that precedes the rest of the
input.

By default, we will assume `false`.

<Playground>
  <InputPhoneNumber region={"NorthAmerica"} showCountryCode={true} />
</Playground>
