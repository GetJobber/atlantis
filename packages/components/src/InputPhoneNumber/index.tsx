import React from "react";
import {
  InputPhoneNumberLegacyProps,
  InputPhoneNumberRebuiltProps,
} from "./InputPhoneNumber.types";
import { InputPhoneNumberRebuilt } from "./InputPhoneNumber.rebuilt";
import { InputPhoneNumber as InputPhoneNumberLegacy } from "./InputPhoneNumber";

export type InputPhoneNumberShimProps =
  | InputPhoneNumberLegacyProps
  | InputPhoneNumberRebuiltProps;

function isNewInputPhoneNumberProps(
  props: InputPhoneNumberShimProps,
): props is InputPhoneNumberRebuiltProps {
  return props.version === 2;
}

export function InputPhoneNumber(props: InputPhoneNumberShimProps) {
  if (isNewInputPhoneNumberProps(props)) {
    return <InputPhoneNumberRebuilt {...props} />;
  } else {
    return <InputPhoneNumberLegacy {...props} />;
  }
}

export { InputPhoneNumberLegacyProps, InputPhoneNumberRebuiltProps };
