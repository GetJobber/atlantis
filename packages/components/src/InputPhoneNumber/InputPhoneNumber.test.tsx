import React from "react";
import renderer from "react-test-renderer";
import { cleanup, fireEvent, render } from "@testing-library/react";
import { InputPhoneNumber } from ".";
import { AllowedRegions } from "./InputPhoneNumber";

afterEach(cleanup);

describe("InputPhoneNumber", () => {
  it("renders a InputPhoneNumber", () => {
    const tree = renderer
      .create(<InputPhoneNumber region={"NorthAmerica"} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe("region", () => {
    describe("when region is NorthAmerica", () => {
      itWillValidateInput("NorthAmerica");

      it("will return the correct phone number with formatting", () => {
        const onChangeHandler = jest.fn();
        const rendered = render(
          <InputPhoneNumber
            region={"NorthAmerica"}
            placeholder="Phone Number"
            onChange={onChangeHandler}
          />,
        );
        const numberInput = rendered.getByLabelText(
          "Phone Number",
        ) as HTMLInputElement;

        fireEvent.change(numberInput, {
          target: { value: "7802424496" },
        });

        expect(onChangeHandler).toHaveBeenCalledWith("(780) 242-4496");
      });

      it("will return the correct phone number with formatting and country code", () => {
        const onChangeHandler = jest.fn();
        const rendered = render(
          <InputPhoneNumber
            region={"NorthAmerica"}
            placeholder="Phone Number"
            showCountryCode={true}
            onChange={onChangeHandler}
          />,
        );
        const numberInput = rendered.getByLabelText(
          "Phone Number",
        ) as HTMLInputElement;

        fireEvent.change(numberInput, {
          target: { value: "7802424496" },
        });

        expect(onChangeHandler).toHaveBeenCalledWith("+1 (780) 242-4496");
      });

      it("will mask the unentered values", () => {
        const rendered = render(
          <InputPhoneNumber
            region={"NorthAmerica"}
            placeholder="Phone Number"
            alwaysShowMask={false}
          />,
        );
        const numberInput = rendered.getByLabelText(
          "Phone Number",
        ) as HTMLInputElement;

        fireEvent.change(numberInput, {
          target: { value: "78024244" },
        });

        expect(numberInput.value).toBe("(780) 242-44__");
      });
    });

    describe("when region is GreatBritain", () => {
      itWillValidateInput("GreatBritain");

      it("will return the correct phone number with formatting", () => {
        const onChangeHandler = jest.fn();
        const rendered = render(
          <InputPhoneNumber
            region={"GreatBritain"}
            placeholder="Phone Number"
            onChange={onChangeHandler}
          />,
        );
        const numberInput = rendered.getByLabelText(
          "Phone Number",
        ) as HTMLInputElement;

        fireEvent.change(numberInput, {
          target: { value: "7802424496" },
        });

        expect(onChangeHandler).toHaveBeenCalledWith("7802424496");
      });

      it("will return the correct phone number with formatting and country code", () => {
        const onChangeHandler = jest.fn();
        const rendered = render(
          <InputPhoneNumber
            region={"GreatBritain"}
            placeholder="Phone Number"
            showCountryCode={true}
            onChange={onChangeHandler}
          />,
        );
        const numberInput = rendered.getByLabelText(
          "Phone Number",
        ) as HTMLInputElement;

        fireEvent.change(numberInput, {
          target: { value: "7802424496" },
        });

        expect(onChangeHandler).toHaveBeenCalledWith("+44 7802424496");
      });
    });

    describe("when region is Unknown", () => {
      itWillValidateInput("Unknown");

      it("will return the correct phone number with formatting", () => {
        const onChangeHandler = jest.fn();
        const rendered = render(
          <InputPhoneNumber
            region={"Unknown"}
            placeholder="Phone Number"
            onChange={onChangeHandler}
          />,
        );
        const numberInput = rendered.getByLabelText(
          "Phone Number",
        ) as HTMLInputElement;

        fireEvent.change(numberInput, {
          target: { value: "7802424496" },
        });

        expect(onChangeHandler).toHaveBeenCalledWith("7802424496");
      });

      it("will return the correct phone number with formatting and country code (no country code for unknown)", () => {
        const onChangeHandler = jest.fn();
        const rendered = render(
          <InputPhoneNumber
            region={"Unknown"}
            placeholder="Phone Number"
            showCountryCode={true}
            onChange={onChangeHandler}
          />,
        );
        const numberInput = rendered.getByLabelText(
          "Phone Number",
        ) as HTMLInputElement;

        fireEvent.change(numberInput, {
          target: { value: "7802424496" },
        });

        expect(onChangeHandler).toHaveBeenCalledWith("7802424496");
      });
    });

    describe("when region is not provided", () => {
      itWillValidateInput(undefined);

      it("will return a NorthAmerican formatting", () => {
        const onChangeHandler = jest.fn();
        const rendered = render(
          <InputPhoneNumber
            placeholder="Phone Number"
            onChange={onChangeHandler}
          />,
        );
        const numberInput = rendered.getByLabelText(
          "Phone Number",
        ) as HTMLInputElement;

        fireEvent.change(numberInput, {
          target: { value: "7802424496" },
        });

        expect(onChangeHandler).toHaveBeenCalledWith("(780) 242-4496");
      });
    });
  });

  describe("alwaysShowMask", () => {
    it("will hide the mask until the user begins to enter valid input", () => {
      const rendered = render(
        <InputPhoneNumber
          region={"NorthAmerica"}
          placeholder="Phone Number"
          alwaysShowMask={false}
        />,
      );
      const numberInput = rendered.getByLabelText(
        "Phone Number",
      ) as HTMLInputElement;

      expect(numberInput.value).toBe("");

      fireEvent.change(numberInput, {
        target: { value: "7802424496" },
      });

      expect(numberInput.value).toBe("(780) 242-4496");
    });
  });
});

function itWillValidateInput(region?: keyof AllowedRegions) {
  it("will only accept numerical characters", () => {
    const onChangeHandler = jest.fn();

    const rendered = render(
      <InputPhoneNumber
        region={region}
        placeholder="Phone Number"
        onChange={onChangeHandler}
      />,
    );
    const numberInput = rendered.getByLabelText(
      "Phone Number",
    ) as HTMLInputElement;

    fireEvent.change(numberInput, {
      target: { value: "abcde~!@#$%^&*()_+={}[]\\|`?/\"'`" },
    });

    expect(onChangeHandler).toHaveBeenCalledTimes(0);
  });
}
