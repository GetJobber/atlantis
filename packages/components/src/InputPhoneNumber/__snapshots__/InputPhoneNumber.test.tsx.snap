// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputPhoneNumber renders a InputPhoneNumber 1`] = `
<div
  className="wrapper"
  style={
    Object {
      "--formField-maxLength": undefined,
    }
  }
>
  <div
    className="inputWrapper"
  >
    <label
      className="label"
      htmlFor="123e4567-e89b-12d3-a456-426655440001"
    />
    <input
      className="input"
      id="123e4567-e89b-12d3-a456-426655440001"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseUp={[Function]}
      type="text"
      value="(___) ___-____"
    />
  </div>
</div>
`;
